diff --git a/vllm/model_executor/models/qwen2.py b/vllm/model_executor/models/qwen2.py
index 54dc0bebd..c154decf2 100644
--- a/vllm/model_executor/models/qwen2.py
+++ b/vllm/model_executor/models/qwen2.py
@@ -163,6 +163,13 @@ class Qwen2Attention(nn.Module):
         )
         attn_cls = (EncoderOnlyAttention
                     if attn_type == AttentionType.ENCODER_ONLY else Attention)
+        
+        self.disable_attn = False
+        import os
+        if "VLLM_DISABLE_ATTN" in os.environ and os.environ["VLLM_DISABLE_ATTN"] == "1":
+            print("Hijacking the attention in Qwen2 to forbid the use of the attention.")
+            self.disable_attn = True
+            
         self.attn = attn_cls(
             self.num_heads,
             self.head_dim,
@@ -185,7 +192,11 @@ class Qwen2Attention(nn.Module):
         qkv, _ = self.qkv_proj(hidden_states)
         q, k, v = qkv.split([self.q_size, self.kv_size, self.kv_size], dim=-1)
         q, k = self.rotary_emb(positions, q, k)
-        attn_output = self.attn(q, k, v)
+        # For profiling, we hijack here to forbid the use of the attention.
+        if self.disable_attn:
+            attn_output = q
+        else:
+            attn_output = self.attn(q, k, v)
         output, _ = self.o_proj(attn_output)
         return output
 
diff --git a/vllm/v1/engine/async_llm.py b/vllm/v1/engine/async_llm.py
index e467f40f6..3e69748ab 100644
--- a/vllm/v1/engine/async_llm.py
+++ b/vllm/v1/engine/async_llm.py
@@ -487,6 +487,16 @@ class AsyncLLM(EngineClient):
 
         self.output_handler = asyncio.create_task(output_handler())
 
+    async def abort_all(self) -> int:
+        """Abort all requests in OutputProcessor and EngineCore."""
+        all_request_ids = self.output_processor.abort_all()
+        await self.engine_core.abort_requests_async(all_request_ids)
+
+        if self.log_requests:
+            logger.info("Aborted all requests %s.", ",".join(all_request_ids))
+        
+        return len(all_request_ids)
+
     async def abort(self, request_id: Union[str, Iterable[str]]) -> None:
         """Abort RequestId in OutputProcessor and EngineCore."""
 
diff --git a/vllm/v1/engine/output_processor.py b/vllm/v1/engine/output_processor.py
index 02c8c61cb..55cd3f3d4 100644
--- a/vllm/v1/engine/output_processor.py
+++ b/vllm/v1/engine/output_processor.py
@@ -311,6 +311,11 @@ class OutputProcessor:
             assert state.queue is not None
             state.queue.put(e)
 
+    def abort_all(self) -> list[str]:
+        abort_request_ids = list(self.request_states.keys())
+        self.abort_requests(abort_request_ids)
+        return abort_request_ids
+
     def abort_requests(
         self,
         request_ids: Iterable[str],
@@ -413,6 +418,8 @@ class OutputProcessor:
             req_state.num_cached_tokens = engine_core_output.num_cached_tokens
             req_state.is_prefilling = False
 
+            # print(f"In output processor, request {req_id} finish_reason is {finish_reason} and stop_reason is {stop_reason}")
+
             if pooling_output is None:
                 assert req_state.detokenizer is not None
                 assert req_state.logprobs_processor is not None
diff --git a/vllm/v1/metrics/loggers.py b/vllm/v1/metrics/loggers.py
index 347185d83..cae85eca9 100644
--- a/vllm/v1/metrics/loggers.py
+++ b/vllm/v1/metrics/loggers.py
@@ -680,8 +680,11 @@ class StatLoggerManager:
                         and issubclass(logger_factory, PrometheusStatLogger)):
                     prometheus_factory = logger_factory
                     continue
-                loggers.append(logger_factory(vllm_config,
-                                              engine_idx))  # type: ignore
+                if not isinstance(logger_factory, Callable):
+                    loggers.append(logger_factory)
+                else:
+                    loggers.append(logger_factory(vllm_config,
+                                                engine_idx))  # type: ignore
             self.per_engine_logger_dict[engine_idx] = loggers
 
         # For Prometheus, need to share the metrics between EngineCores.
diff --git a/vllm/v1/metrics/stats.py b/vllm/v1/metrics/stats.py
index e6c344d19..c44f32476 100644
--- a/vllm/v1/metrics/stats.py
+++ b/vllm/v1/metrics/stats.py
@@ -30,6 +30,11 @@ class PrefixCacheStats:
 class SchedulerStats:
     """Stats associated with the scheduler."""
 
+    # NOTE(lhy): we need to add this field to the scheduler stats for PSRL.
+    need_to_abort_reqs: list[str] = field(default_factory=list)
+    req_id_to_prompt_token_num: dict[str, int] | None = None
+    req_id_to_response_token_num: dict[str, int] | None = None
+    
     num_running_reqs: int = 0
     num_waiting_reqs: int = 0
 
